Feature/cldsrv 322 introduce sosapi internal routes

williamlardier opened the pull request:
Context Veeam12 SOSAPI support. See the commit descriptions for more information.

This PR introduces a new set of APIs exposed as internal routes for veeam. The ZKOP PR handling redirection is open here: https://github.com/scality/zenko-operator/pull/355.

The high level approach of this pr is:

* Create a new router that will manbage the HTTP verbs, authentication + authorization, and then call the right API based on the input parameters (or return an error).
* A utils file to handle common operations like streaming data, or getting streamed data from the client.
* One file per API: PUT, GET, DELETE, HEAD and LIST, with support for Veeam12 files, to be stored in the bucket metadata. * The associated bucket metadata changes are done in https://github.com/scality/Arsenal/pull/2059.
* XML schema validation for Veeam12 files. If the protocol version is known, the schema is validated, otherwise, any valid XML file is accepted, with a 1MB size limit.
* This implementation covers UI, API and CLI usage.

Please let me know if you have any question on the context.

This was tested end2end in a real Artesca with the ZKOP changes:

## This was tested end2end along with Data Metrics and ZKOP custom Ingresses with success:

![image](https://user-images.githubusercontent.com/91462779/210088640-f6a54ea8-73af-4e29-bb99-896f337946ab.png)

![image](https://user-images.githubusercontent.com/91462779/210367506-61bf1540-0f7e-4a98-aac2-81b936fedda7.png)

![image](https://user-images.githubusercontent.com/91462779/210366856-d1b77883-2631-4900-8838-2330f262b9a9.png)

bert-e commented:
# Hello williamlardier,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-322 contains:


* `8.7.2`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.7.3`


Please check the `Fix Version/s` of CLDSRV-322, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

anurag4DSB reviewed (commented):
Would it be possible to these into smaller logical PRs?
Its currently very long to review.
We can piggyback branches if needed.

Seragonia committed:
CLDSRV-322: bump arsenal version

Seragonia committed:
CLDSRV-322: Create a new route for Veeam12 SOS API.

This new route is exposed through special nginx rules
from Zenko-Operator, to redirect any call to the veeam
folder, located under .system-d26a9498-cb7c-4a87-a44a-8ae204f5ba6c
at the root of the bucket. The goal is to store files in
the bucket metadata, to ease their update by internal jobs.
To avoid impacting standard API, we rely on custom routes
with dedicated logic to handle these files, in a generic
way.

This commit introduces a new route that will manage, in turn,
the:
- Handling of incoming request.
- Validity checks, including list of suppoorted APIs according
  to the HTTP verb and query parameters.
- Authentication and Authorization with Vault, in the same
  way as usual files.
- Check of the targeted bucket and/or keys, to extract the
  bucket metadata.
- Routing of the request to the right API handler.

Seragonia committed:
CLDSRV-322: implement SOSAPI scheme validator

SOSAPI relies on standard XML files for both the system and the capacity.
It is used by Veeam12+ to determine what capabilities and/or
configuration should be enforced for a given S3-integrated Bucket used
for backups.

The commit introduces scheme validation for JSON objects, as XML will
be first converted using xml2js.

The system.xml file includes the protocol version of SOSAPI: if the
version is not know, no validation is made, to allow for future changes
without formal need to update the product.

Note: maximum XML file size, in case of unsupported protocol version, will
be enforced to avoid spacing issues with the database.

Seragonia committed:
CLDSRV-322: implement common util functions

Custom SOSAPI routes might either retreive or stream data. The utils file
re-implement, with support for this particular context, some functions
from the standard API paths, from Arsenal.

These changes mostly introduce ways to compute the right HTTP headers as
well as input our output streams to handle GET or PUT request types.

Seragonia committed:
CLDSRV-322: Implement GET for SOSAPI

The GET method is used by SOSAPI to determine if SOS API is enabled
or not on a bucket.

Two files are supported: system.xml and capacity.xml.

This API:
- Get the bucket metadata
- Dynamically recomputes a valid XML based on the bucket md content
  using xml2js as headless, to enforce the same XML as the one
  from SOSAPI standard
- Rejects the request with an error if the bucket metadata does not
  exist
- Handle the `?tagging` request, required for versioned bucket, to
  return a static content.

Output stream relies on the utils file.

Seragonia committed:
CLDSRV-322: Implement PUT for SOSAPI routes

In the SOSAPI context, the user is requested to pre-created two files,
system.xml and capacity.xml under the veeam folder, to enable the feature.

This API:
- Extracts the XML from the provided file, convert it to JSON
- Validate that the JSON is valid against joi schemes, if applicable
- Updates the bucket metadata, including the last-modified date
- Update in the database the bucket metadata
- Return the standard success code response
- Handle invalid XML or XML structure, and return an error accordingly

Seragonia committed:
CLDSRV-322: Implement HEAD for SOSAPI routes

HEAD object is not formally required by Veeam SOSAPI, but Veeam
relies on the last-modified date value of the capacity.xml file.
To suppoort any change in future SOSAPI standard, the HEAD method
is implemented, and is similar to the GET method, where only the metadata
are returned.

Seragonia committed:
CLDSRV-322: Implement DELETE for SOSAPI routes

Deletion of Veam SOSAPI files are required to ensure consistent
user experience. This API is only exposed to API/CLI clients, not
User Interface. The User Interface relies on MultiObjectDelete to
perform the deletions, and is handled in a separate commit.

This API:
- Checks that the requested file exists
- Erase the bucket metadata according to the file
- Update the bucket metadata with the updated values.
- Handle errors if the Veaam capability is not yet enabled for the bucket

Seragonia committed:
CLDSRV-322: Implement LIST for SOSAPI routes

Listing of objects is needed for consistent user experience in the
product's User Interface.

Listing is implemented as a `GET` request with a specific query parameter
`list-type` and folder `.system-d26a9498-cb7c-4a87-a44a-8ae204f5ba6c`.

This API:
- Handles both versioned and non-versioned listing
- Relies on predefined templates to fill the response content
- Extracts the system.xml and capacity.xml files from the bucket
  metadata
- Computes the listing response based on the input query parameters
  and files currently in the bucket md capabilities
- Handle errors if any non supported query parameter is used. As any GET
  request is routed to this method, we return InvalidRequest if the requested
  action is not supported (i.e., not a listing v2)

Seragonia committed:
CLDSRV-322: Support custom files for MultiObjectDelete

MultiObjectDelete is implemented by the product UI to delete the
files in buckets. This method is a POST that relies on the request
body to filter the objects, hence, it is not possible to filter
it as an ingress rule in nginx.

The implementation tries to avoid adding any complexity
by extending existing loops, and implementing a new step if elligible
files are found.

These files are extracted from the Veeam route list of accepted files,
but this implementation might change if more custom APIs are supported
in the future.

Seragonia committed:
CLDSRV-322: Implement test for custom routes

Unit and funcitonal tests are implemented to test the custom routes.
The LISTing is not yet tested, as it requires more changes to
generate a valid signature, from Mocha.

Seragonia committed:
CLDSRV-322: bump cloudserver version

francoisferrand commented on package.json:
```diff
@@ -21,7 +21,7 @@
     "@azure/storage-blob": "^12.12.0",
     "@hapi/joi": "^17.1.0",
-    "arsenal": "git+https://github.com/scality/Arsenal#8.1.77",
+    "arsenal": "git+https://github.com/scality/Arsenal#f446354c526384dc93bd3d2c79c39e50b1a8186c",
```
please do not forget to update before merging

francoisferrand commented on lib/routes/routeVeeam.js:
```diff
@@ -0,0 +1,219 @@
+ * @returns {Error | undefined} error or undefined
+ */
+function checkBucketAndKey(bucketName, objectKey, requestQueryParams, method, log) {
+    if (!bucketName && !(method === 'GET' && !objectKey)) {
```
kind of hard to read (too many negations), maybe consider rewriting in a way that would be easier to read... (e.g. introducing helper function, to "name" the intent; or splitting in multiple `if` statements)

williamlardier replied:
This was coming from Arsenal, see example here: https://github.com/scality/Arsenal/blob/development/8.1/lib/s3routes/routes.ts#L48

I will change the logic to make it more easy to read.

francoisferrand replied:
much more readable indeed (esp. with the comment 😀), but missing the `else` to behave the same?

williamlardier replied:
Actually the return should be in the nested if, no need for else here, and then we will achieve the same result

williamlardier replied:
I updated witht he old way of checking with a comment in https://github.com/scality/cloudserver/pull/4942/commits/93776d6f12455799f93951ef3da1753303db1439#diff-d54eeee0f403b38d9cec6ec8ab8caca914402b5705e4bfb96179c5065b8ad8a0R51, as per discussion over Slack

francoisferrand commented on lib/routes/routeVeeam.js:
```diff
@@ -0,0 +1,224 @@
+                requestType,
+                request,
+            };
+            return metadataValidateBucket(mdValParams, log, (err, _bucketMd) => {
```
could use waterfall instead, to avoid nesting & uneeded error checking:

```js
    return metadataValidateBucket(mdValParams, log, next)
},
(_bucketMd, next) => {
    bucketMd = _bucketMd;
    return next();
});
```

or better yet, remove the bucketMd variable and just pass it to the final callack:
```js
return async.waterfall([
    ...
    (userInfo, next) => {
         ...
         return metadataValidateBucket(mdValParams, log, next)
    }
], (err, bucketMd) {
    if (err || !bucketMd) {
        return responseXMLBody(err, null, response, log);
    }
    return callback(request, response, bucketMd, log);
});
```

williamlardier replied:
Done here: https://github.com/scality/cloudserver/pull/4942/commits/3941ed635c72ef5bf941e7a729d681a09403b525#diff-d54eeee0f403b38d9cec6ec8ab8caca914402b5705e4bfb96179c5065b8ad8a0R130

francoisferrand commented on lib/routes/routeVeeam.js:
```diff
@@ -0,0 +1,224 @@
+    });
+}
+
+function _normalizeVeeamRequest(req) {
```
do we really _have to_ replace some fields from the req?

Apart from the (not yet relevant) potential type-ability implications it may have, this kind of pattern may leave the object in a partially updated state, and lead to hard to find issues : it is generally safer to create a new normalized object with exactly the normalizedinformation you need...

williamlardier replied:
Yes, the url must be updated to account for the dynamic path change from the ingress, in order to have valid V4 signatures.
This is actually something that we already do in Arsenal and other custom routes as well, I am not sure we want to change the pattern in only one place, do we?

Example from Arsenal: https://github.com/scality/Arsenal/blob/development/8.1/lib/s3routes/routesUtils.ts#L1039

francoisferrand replied:
probably not now indeed... 😞

francoisferrand commented on lib/routes/veeam/get.js:
```diff
@@ -0,0 +1,58 @@
+            delete fileToBuild.CapacityInfo?.LastModified;
+        }
+        modified = fileToBuild.LastModified;
+        delete fileToBuild.LastModified;
```
this is never set. fileToBuild only has a single field named either `SystemInfo` or `CapacityInfo`

williamlardier replied:
I created some util functions that now better handle this: https://github.com/scality/cloudserver/pull/4942/commits/3941ed635c72ef5bf941e7a729d681a09403b525#diff-a23b06e1deac8b9332f76c3ae6723a32d6a49e7b7845e7f94f76e17190be0eb0R153

francoisferrand commented on lib/routes/veeam/head.js:
```diff
@@ -0,0 +1,44 @@
+        }
+        const isSystemXML = request.objectKey.endsWith('/system.xml');
+        const fileToBuild = isSystemXML ? data._capabilities?.VeeamSOSApi?.SystemInfo
+            : isSystemXML && !data._capabilities?.VeeamSOSApi?.CapacityInfo;
```
the 'else' branch will always evaluate to `false`, since it includes the 'root' condition...

williamlardier replied:
Updated in https://github.com/scality/cloudserver/pull/4942/commits/3941ed635c72ef5bf941e7a729d681a09403b525#diff-ee4e7989b66eaaa0e7f62d59ec42d4cfa0d5c6a8bf5942a69b07110aea5555d1R24 (it relies on the new util functions, that solves the issue)

francoisferrand commented on lib/routes/veeam/head.js:
```diff
@@ -0,0 +1,44 @@
+            return responseXMLBody(errors.NoSuchKey, null, response, log);
+        } else if (!isSystemXML && !data._capabilities?.VeeamSOSApi?.CapacityInfo) {
+            return responseXMLBody(errors.NoSuchKey, null, response, log);
+        }
```
```js
if (!fleToBuild) {
    return responseXMLBody(errors.NoSuchKey, null, response, log);
}
```

williamlardier replied:
Done in https://github.com/scality/cloudserver/pull/4942/commits/3941ed635c72ef5bf941e7a729d681a09403b525#diff-a23b06e1deac8b9332f76c3ae6723a32d6a49e7b7845e7f94f76e17190be0eb0R174

francoisferrand commented on lib/routes/veeam/list.js:
```diff
@@ -0,0 +1,162 @@
+        baseXML.push(`<LastModified>${file['Last-Modified']}</LastModified>`);
+        baseXML.push(`<ETag>${file.ETag}</ETag>`);
+        baseXML.push(`<Size>${file['Content-Length']}</Size>`);
+        baseXML.push('<StorageClass>STANDARD</StorageClass>');
```
is the `storageClass` value required by Veam?
Otherwise, could we (and should we?) use some specific value, to better reflect this is a veam-specific "virtual" object ?

williamlardier replied:
This is not required, and you are right, it would make more sense, I'll change it, thanks.

williamlardier replied:
I tried with VIRTUAL, let me know what you think: https://github.com/scality/cloudserver/pull/4942/commits/3941ed635c72ef5bf941e7a729d681a09403b525#diff-25506d7604624dfc2c3d66729bd5dffc4534b63a3bffd2d7213ffa1dcb794ae5R47

![image](https://user-images.githubusercontent.com/91462779/210557451-2e465117-73de-4796-aae4-1775f11cc854.png)

francoisferrand replied:
i guess that is a question for PO / Product Director...

williamlardier replied:
I will ask Thomas then, thanks!

williamlardier replied:
He agrees on the VIRTUAL name, to be documented in the SOSAPI documentation for Artesca.

francoisferrand commented on lib/routes/veeam/list.js:
```diff
@@ -0,0 +1,162 @@
+const { responseXMLBody } = require('arsenal/build/lib/s3routes/routesUtils');
+const { respondWithData, getResponseHeader, buildHeadXML, validPath } = require('./utils');
+
+function buildContent(arrayOfFiles) {
```
instead of rewriting these XML-generating functions, I think we should reuse the existing `processVersions()`, `processMasterVersions()`, [...] functions (from `bucketGet.js`): e.g. here we would just build the proper parameter to these functions.

It would be less redundant, and allow updating the xml management stuff in a single place (and ideally these API will be typed, at some point...)

williamlardier replied:
I did what you suggested, so I created a function to create the right objects before calling the usual functions from `bucketGet`: https://github.com/scality/cloudserver/pull/4942/commits/3941ed635c72ef5bf941e7a729d681a09403b525#diff-25506d7604624dfc2c3d66729bd5dffc4534b63a3bffd2d7213ffa1dcb794ae5R24

Do you think the object template should be made in a separate function? I find it not really helpful here, so I'm repeating it once for the base file (the folder).

williamlardier replied:
List response looks like this (tested with Artesca):

![image](https://user-images.githubusercontent.com/91462779/210558056-994167e8-e5f3-4c89-9034-b72dcebc2733.png)

williamlardier replied:
Note: the double-double quotes for the ETags are now solved in https://github.com/scality/cloudserver/pull/4942/commits/067383064e138bb79da293d36b6189bd97ed3ee1#diff-25506d7604624dfc2c3d66729bd5dffc4534b63a3bffd2d7213ffa1dcb794ae5R42

francoisferrand commented on lib/routes/veeam/head.js:
```diff
@@ -0,0 +1,44 @@
+        if (err) {
+            return responseXMLBody(errors.InternalError, null, response, log);
+        }
+        const isSystemXML = request.objectKey.endsWith('/system.xml');
+        const fileToBuild = isSystemXML ? data._capabilities?.VeeamSOSApi?.SystemInfo
+            : isSystemXML && !data._capabilities?.VeeamSOSApi?.CapacityInfo;
+        if (isSystemXML && !data._capabilities?.VeeamSOSApi?.SystemInfo) {
+            return responseXMLBody(errors.NoSuchKey, null, response, log);
+        } else if (!isSystemXML && !data._capabilities?.VeeamSOSApi?.CapacityInfo) {
+            return responseXMLBody(errors.NoSuchKey, null, response, log);
+        }
+        const savedModifiedDate = fileToBuild.LastModified;
+        delete fileToBuild.LastModified;
+        // Recompute file content to generate appropriate content-md5 header
+        const builder = new xml2js.Builder({
+            headless: true,
+        });
```
this whole mapping of `system.xml` to `SystemInfo` and `CapacityInfo` otherwise to generate the XML is duplicate in both Head and Get, as well as list to some extend : should be factorized/de-duplicated (for get & head definitely, not sure how it would go for list...)

williamlardier replied:
see https://github.com/scality/cloudserver/pull/4942/commits/3941ed635c72ef5bf941e7a729d681a09403b525#diff-a23b06e1deac8b9332f76c3ae6723a32d6a49e7b7845e7f94f76e17190be0eb0R159

williamlardier replied:
Note: I squashed again to update the Arsenal version used and some linter issues. See https://github.com/scality/cloudserver/pull/4942/commits/067383064e138bb79da293d36b6189bd97ed3ee1

anurag4DSB commented:
> Would it be possible to these into smaller logical PRs?
> 
> Its currently very long to review.
> 
> We can piggyback branches if needed.

Post discussion with @francoisferrand  one PR is good

Seragonia committed:
CLDSRV-322: code improvements

anurag4DSB commented on lib/routes/veeam/utils.js:
```diff
@@ -0,0 +1,16 @@
+    return decodeURIComponent(uri.replace(/\+/g, ' '));
+}
+
+const validPath = '.system-d26a9498-cb7c-4a87-a44a-8ae204f5ba6c/';
```
would it be better to add this in https://github.com/scality/cloudserver/blob/development/8.6/constants.js?

williamlardier replied:
I understand but I feel like we shouldn't mix custom route and standard configuration... And at some point, it may even be set at the kubernetes level... I'm fine putting it in constants.js but for now I do not really see a benefit of doing so. Please feel free to challenge my point of view

anurag4DSB commented on lib/routes/routeVeeam.js:
```diff
@@ -0,0 +1,219 @@
+ */
+function authorizationMiddleware(request, response, api, log, callback) {
+    if (!api) {
+        return errors.AccessDenied;
```
Should we return callback here?

williamlardier replied:
Actually it should return a call to `responseXMLBody` with this error, the callback is the API

williamlardier replied:
Done: https://github.com/scality/cloudserver/pull/4942/commits/e31d948aad0278ca1d2994d0999f573c437ba724#diff-d54eeee0f403b38d9cec6ec8ab8caca914402b5705e4bfb96179c5065b8ad8a0R96

anurag4DSB commented on lib/routes/routeVeeam.js:
```diff
@@ -0,0 +1,219 @@
+                    bucketName: request.bucketName,
+                    objectKey: request.objectKey,
+                });
+            }
```
nit: identation

williamlardier replied:
All good in https://github.com/scality/cloudserver/pull/4942/files#diff-d54eeee0f403b38d9cec6ec8ab8caca914402b5705e4bfb96179c5065b8ad8a0R101, my squash commits will fix it in the commit you referenced.

anurag4DSB commented on lib/routes/routeVeeam.js:
```diff
@@ -0,0 +1,219 @@
+
+function _normalizeVeeamRequest(req) {
+    /* eslint-disable no-param-reassign */
+    // Rewriting the URL is needed for the V4 signature check
```
Would be great to mention also, why, rewriting the URL is needed for the V4 signature check

williamlardier replied:
I hope this is clear enough: https://github.com/scality/cloudserver/pull/4942/commits/e31d948aad0278ca1d2994d0999f573c437ba724#diff-d54eeee0f403b38d9cec6ec8ab8caca914402b5705e4bfb96179c5065b8ad8a0R141

williamlardier commented on lib/routes/veeam/list.js:
```diff
@@ -32,7 +32,7 @@ function buildXMLResponse(request, arrayOfFiles, versioned = false) {
         CommonPrefixes: [],
     };
-    arrayOfFiles.forEach(file => list[versioned ? 'Versions' : 'Contents'].push({
+    const entries = arrayOfFiles.map(file => ({
```
Just to keep track of your comments @francoisferrand :

> may be written more expressively like this:

```
const entries = arrayOfFiles.map(file => { ..... })
// Add the folder as the base file
entries.push({ ... }
```

> and that 'same' list of entries can be used to build the appropriate `list` object afterwards, without the error-prone `list[versioned ? 'Versions' : 'Contents']`


-> It's handled here https://github.com/scality/cloudserver/pull/4942/commits/93776d6f12455799f93951ef3da1753303db1439#diff-25506d7604624dfc2c3d66729bd5dffc4534b63a3bffd2d7213ffa1dcb794ae5R35
ALso tested with Artesca

anurag4DSB reviewed (approved):
Logic LGTM per design + extra enhancements for the UI.

Small comments, and mostly look good with all the tests.
I have reviewed the arsenal PR but have yet to review the ZKOP one.

francoisferrand commented on lib/routes/veeam/put.js:
```diff
@@ -0,0 +1,90 @@
+            }
+            // Validate the JS object schema with joi and prepare the object for
+            // further logic
+            const validateFn = isSystemXML ? parseSystemSchema : parseCapacitySchema;
```
can use `isSystemXML()`

williamlardier replied:
Done: https://github.com/scality/cloudserver/pull/4942/commits/8c671ef0ac52fb1f0100cbc68a7f5da0eb9388d5#diff-38a3a2f10ad2b2920ef0e2d5815a184a4bdaab48fe7c82ecc0767c0be2fb8c8aR50

francoisferrand commented on lib/routes/veeam/put.js:
```diff
@@ -0,0 +1,90 @@
+            }
+            // System and Capacity files also include the last-modified date used by Veeam
+            const lastModified = (new Date()).toISOString();
+            if (isSystemXML) {
```
we may actually be able to use `getFileToBuild()` here from utils ?

something like :

```js
capabilities.VeeamSOSApi = getFileToBuild(request, validatedData).value
```

williamlardier replied:
Done with error handling: https://github.com/scality/cloudserver/pull/4942/commits/8c671ef0ac52fb1f0100cbc68a7f5da0eb9388d5#diff-38a3a2f10ad2b2920ef0e2d5815a184a4bdaab48fe7c82ecc0767c0be2fb8c8aR58

Also, I cannot set to capabilities directly, it would override existing data, so I'm just overriding if needed.

francoisferrand commented on lib/api/multiObjectDelete.js:
```diff
@@ -481,6 +484,9 @@ function multiObjectDelete(authInfo, request, log, callback) {
                         if (result.isAllowed) {
+                            if (validObjectKeys.includes(entry.key)) {
+                                inPlayInternal.push(entry.key);
+                            }
```
i am wondering about this.

When the object is a "veam" object, we add it to "inPlayInteral" so we will process it with veam-specific code (which is fine), but we also add it to inPlay:
* so we will run checkBucketMetadata, which is needed i guess?
* eventually we will call getObjMetadataAndDeleteStep() : so we will try to delete a "real" object with that name from the "real" metadata backend? which will likely fail (since it would not be present)?

Or am I missing something?

williamlardier replied:
You are right, I moved the inPlay in the `else` to make sure we do not duplicate the custom files
See https://github.com/scality/cloudserver/pull/4942/commits/8c671ef0ac52fb1f0100cbc68a7f5da0eb9388d5#diff-e8319ec288d282012f389d259d6468a0922b91ec7cc52f2b8c15b01ff3ba9c2aR489

Without this change, there is no bad effects, it just do more computations than needed.

francoisferrand commented on lib/routes/veeam/put.js:
```diff
@@ -0,0 +1,90 @@
+            return next(null, parsed);
+        }),
+        (parsedXML, next) => {
+            const isSystemXML = request.objectKey.endsWith('/system.xml');
```
isSystemXML()

williamlardier replied:
See https://github.com/scality/cloudserver/pull/4942/commits/8c671ef0ac52fb1f0100cbc68a7f5da0eb9388d5#diff-38a3a2f10ad2b2920ef0e2d5815a184a4bdaab48fe7c82ecc0767c0be2fb8c8aR50

francoisferrand commented on lib/routes/veeam/delete.js:
```diff
@@ -0,0 +1,73 @@
+ * @returns {undefined} -
+ */
+function deleteVeeamCapabilities(bucketName, objectKey, bucketMd, log, callback) {
+    const isSystemXML = objectKey.endsWith('system.xml');
```
isSystemXML()

williamlardier replied:
See https://github.com/scality/cloudserver/pull/4942/commits/8c671ef0ac52fb1f0100cbc68a7f5da0eb9388d5#diff-3d01b0ce4989d0b3f542dc7de83f55537d4fbfa70d7d63258b7201b7dffdd95bR18

francoisferrand commented on lib/routes/veeam/put.js:
```diff
@@ -0,0 +1,83 @@
+        }),
+        (parsedXML, next) => {
+            const capabilities = bucketMd._capabilities || {
+                VeeamSOSApi: {},
+            };
+            // Initialize Veeam capability if not already there
+            if (!capabilities.VeeamSOSApi) {
+                capabilities.VeeamSOSApi = {};
+            }
```
not really needed:
* the 'default' initi is redundant with the later 'Initialize Veeam capability if not already there'
* all these are actually not needed, since we set it anyway line 62: ` capabilities.VeeamSOSApi = { ...`

i guess we could simply do this (at line 62):
```
            capabilities.VeeamSOSApi = {
                ...(capabilities.VeeamSOSApi || {}),
                ...file.value,
            };
```

ghost commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.7`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.4`
* `development/8.5`
* `development/8.6`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.7`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.4`
* `development/8.5`
* `development/8.6`



Please check the status of the associated issue CLDSRV-322.

Goodbye williamlardier.
